api/task_manager.cc

```diff
@@ -0,0 +1,165 @@
+    res.type = status.task_status.type;
+    res.state = status.task_status.state;
+    res.is_abortable = bool(status.abortable);
+    res.start_time = st;
+    res.end_time = et;
+    res.error = status.task_status.error;
+    res.parent_id = status.parent_id.to_sstring();
+    res.sequence_number = status.task_status.sequence_number;
+    res.shard = status.task_status.shard;
+    res.keyspace = status.task_status.keyspace;
+    res.table = status.task_status.table;
+    res.entity = status.task_status.entity;
+    res.progress_units = status.task_status.progress_units;
+    res.progress_total = status.progress.total;
+    res.progress_completed = status.progress.completed;
+    return res;
+}
+
+future<json::json_return_type> retrieve_status(task_manager::foreign_task_ptr task) {
+    if (task.get() == nullptr) {
+        co_return coroutine::return_exception(httpd::bad_param_exception("Task not found"));
+    }
+    auto progress = co_await task->get_progress();
+    full_task_status s;
+    s.task_status = task->get_status();
+    s.parent_id = task->get_parent_id();
+    s.abortable = task->is_abortable();
+    s.module = task->get_module()->get_name();
+    s.progress.completed = progress.completed;
+    s.progress.total = progress.total;
+    co_return make_status(s);
+}
+
+void set_task_manager(http_context& ctx, routes& r) {
+    tm::get_modules.set(r, [&ctx] (std::unique_ptr<request> req) -> future<json::json_return_type> {
+        std::vector<std::string> v = boost::copy_range<std::vector<std::string>>(ctx.tm.local().get_modules() | boost::adaptors::map_keys);
+        co_return v;
+    });
+
+    tm::get_tasks.set(r, [&ctx] (std::unique_ptr<request> req) -> future<json::json_return_type> {
+        using chunked_stats = utils::chunked_vector<task_stats>;
+        std::vector<chunked_stats> res = co_await ctx.tm.map([&req] (task_manager& tm) {
+            chunked_stats local_res;
+            auto module = tm.find_module(req->param["module"]);
+            const auto& filtered_tasks = module->get_tasks() | boost::adaptors::filtered([&params = req->query_parameters] (const auto& task) {
+                return filter_tasks(task.second, params);
+            });
+            for (auto& [task_id, task] : filtered_tasks) {
+                task_stats ts;
+                ts.task_id = task->id().to_sstring();
+                ts.state = task->get_status().state;
+                local_res.push_back(ts);
+            }
+            return local_res;
+        });
+
+        std::function<future<>(output_stream<char>&&)> f = [r = std::move(res)] (output_stream<char>&& s) -> future<> {
+            bool first = true;
+            auto res = std::move(r);
+            co_await s.write("[");
+            for (auto& v: res) {
+                for (auto& stats: v) {
+                    auto fut = first ? make_ready_future<>() : s.write(", ");
+                    first = false;
+                    co_await std::move(fut);
```

Comment:
This way `make_ready_future<>()` and `s.write(", ")` would execute immediately. It can be changed like this:
```
auto write_delim = std::function([] () mutable { return make_ready_future<>(); });
...
        co_await std::exchange(write_delim, [&s] () mutable { return s.write(", "); })();
```
but I am not sure if it's worth it.

Reply:
another option is to initialize `delim` to an empty string, write it unconditionally and then set it to `", "`.
This way you don't need first and you have no branches.